### PR TITLE
Improve discount system

### PR DESCRIPTION
**Pull request by PubliAlex**
I created a new PR that follow the PR #853 commited to the latest develop branch and that fix the issue that let a 0$ discount to apply.

@tobias-kuendig I let you test it and let me know if it's ok for you

### CODE REVIEW

**Comment by alex-mokkas** [comment]
Any progress on that?

**Comment by tobias-kuendig** [comment]
Thank you @PubliAlex! I'll review the changes in the coming days.

File: classes/cart/DiscountApplier.php [hunk 1]
```diff
@@ -64,15 +64,19 @@ public function apply(Discount $discount): ?bool
...
             $this->reducedTotal -= $savings;
         }
 
         if ($discount->type === 'rate') {
-            $savings            = $this->total * ($discount->rate / 100);
+            $savings            = $this->calculateSaving($discount);
             $this->reducedTotal -= $savings;
         }
 
+        if($savings == 0) {
```
Comment (tobias-kuendig) [inline classes/cart/DiscountApplier.php:76]: Is it right, that these are the only added lines compared to #853?

**Comment by PubliAlex** [comment]
Yes, and I also made some modifications about database tables cause I saw you also made some change.

The line that check the saving value prevent to display a line with a discount of 0.

But, after some more test, it does not cover all the case. When you set a discount coupon code that apply to a specific product, and want to apply this discount to a cart that don't have this product, you want to display a flash or validation message that say that this discount does not apply to this cart.

Actually, you just don't have any feedback and it try to apply a $0 saving discount, and so, as it's 0, it does not apply.

The problem is that you can't set error message in the discountApplier class, so the code need to be improved to also calculate if the discount can apply in the Discounts trait, but I didn't had time to investigate this, have a short deadline to respect actually.

But in all case where you apply discount without coupon code, it works pretty well.

**Pull request closed without merging**